src/backend/src/services/projects.services.ts

```diff
@@ -0,0 +1,394 @@
+import { isProject, Project, WbsNumber, wbsPipe } from 'shared';
+import projectQueryArgs from '../prisma-query-args/projects.query-args';
+import prisma from '../prisma/prisma';
+import projectTransformer from '../transformers/projects.transformer';
+import { validateChangeRequestAccepted } from '../utils/change-requests.utils';
+import { AccessDeniedException, HttpException, NotFoundException } from '../utils/errors.utils';
+import {
+  addDescriptionBullets,
+  createChangeJsonNonList,
+  createRulesChangesJson,
+  editDescriptionBullets,
+  getHighestProjectNumber,
+  getUserFullName
+} from '../utils/projects.utils';
+import { descBulletConverter, wbsNumOf } from '../utils/utils';
+import { createDescriptionBulletChangesJson } from '../utils/work-packages.utils';
+
+export default class ProjectsService {
+  /**
+   * Get all the projects in the database.
+   * @returns all the projects
+   */
+  static async getAllProjects(): Promise<Project[]> {
+    const projects = await prisma.project.findMany({ where: { wbsElement: { dateDeleted: null } }, ...projectQueryArgs });
+    return projects.map(projectTransformer);
+  }
+
+  /**
+   * Get a single project
+   * @param wbsNumber the wbsNumber of the project to get
+   * @returns the request project
+   * @throws if the wbsNumber is invalid, the project is not found, or the project is deleted
+   */
+  static async getSingleProject(wbsNumber: WbsNumber): Promise<Project> {
+    if (!isProject(wbsNumber)) throw new HttpException(400, `${wbsPipe(wbsNumber)} is not a valid project WBS #!`);
+
+    const { carNumber, projectNumber, workPackageNumber } = wbsNumber;
+
+    const project = await prisma.project.findFirst({
+      where: {
+        wbsElement: {
+          carNumber,
+          projectNumber,
+          workPackageNumber
+        }
+      },
+      ...projectQueryArgs
+    });
+
+    if (!project) throw new NotFoundException('Project', wbsPipe(wbsNumber));
+    if (project.wbsElement.dateDeleted) throw new HttpException(400, 'This project has been deleted!');
+
+    return projectTransformer(project);
+  }
+
+  /**
+   * Create a new project with the given information.
+   * @param user the user creating the project
+   * @param crId the change request id being used to create the project
+   * @param carNumber the car number of the new project
+   * @param name the name of the new project
+   * @param summary the summary of the new project
+   * @param teamId the teamId of the new project
+   * @returns the created project
+   * @throws if the user doesn't have permission or if the change request is invalid
+   */
+  static async newProject(
+    user: User,
+    crId: number,
+    carNumber: number,
+    name: string,
+    summary: string,
+    teamId: string | undefined
+  ): Promise<WbsNumber> {
+    if (user.role === Role.GUEST) throw new AccessDeniedException('Guests cannot create projects');
+
+    await validateChangeRequestAccepted(crId);
+
+    if (teamId) {
```

Comment:
Lines 81-83 here are exactly the same as lines 48-53 in the teams.services file, not sure if this counts as duplicate code that we should try to eliminate. Also, on lines 48-53 in the teams.services file, ...teamQueryArgs were passed in as well. Maybe we should do that too?

<img width="410" alt="Screen Shot 2023-01-24 at 10 17 57 PM" src="https://user-images.githubusercontent.com/60333735/214472554-8c09401b-48e5-4635-9dee-8a14caa498ae.png">

Reply:
i wouldn't count this as duplicated since there's no logic involved, it's just getting something from the database and making sure it exists.

on the other hand, i guess we could abstract ALL of that out into helper functions that throw errors if they don't find it? could be nice........